client/lib/use-local-storage/index.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * External dependencies
+ */
+import { useCallback, useRef, useState } from 'react';
+
+export function useLocalStorage< T >(
+	storageKey: string,
+	initialValue: T | ( () => T )
+): [ T, ( newValue: T | ( ( oldValue: T ) => T ) ) => void ] {
+	const [ storedValue, setStoredValue ] = useState( () => {
+		const item = typeof window === 'undefined' ? null : window.localStorage.getItem( storageKey );
+		if ( item !== null ) {
+			try {
+				return JSON.parse( item ) as T;
+			} catch {
+				// Invalid data stored in localStorage, return initialValue
+			}
+		}
+
+		return initialValue instanceof Function ? initialValue() : initialValue;
```

Comment:
I think the more idiomatic check for function-ness is to use `typeof` 🤔

Reply:
You're probably right. Now that I think of it, reading `instanceof` causes a bit of mental overhead for me as I dig deep into my memory banks of how the prototype chain works in JS, which I hardly have to think about. `typeof` is simpler, for me at least.